Sources/Purchasing/ReceiptFetcher.swift

```diff
@@ -19,26 +19,39 @@ class ReceiptFetcher {
     private let requestFetcher: StoreKitRequestFetcher
     private let receiptParser: ReceiptParser
     private let fileReader: FileReader
+    private let clock: ClockType
+
+    private let lastReceiptRefreshRequest: Atomic<Date?> = nil
 
     let systemInfo: SystemInfo
 
     init(
         requestFetcher: StoreKitRequestFetcher,
         systemInfo: SystemInfo,
         receiptParser: ReceiptParser = .default,
-        fileReader: FileReader = DefaultFileReader()
+        fileReader: FileReader = DefaultFileReader(),
+        clock: ClockType = Clock.default
     ) {
         self.requestFetcher = requestFetcher
         self.systemInfo = systemInfo
         self.receiptParser = receiptParser
         self.fileReader = fileReader
+        self.clock = clock
     }
 
     func receiptData(refreshPolicy: ReceiptRefreshPolicy, completion: @escaping (Data?) -> Void) {
         switch refreshPolicy {
         case .always:
-            Logger.debug(Strings.receipt.force_refreshing_receipt)
-            self.refreshReceipt(completion)
+            if self.shouldThrottleRefreshRequest() {
+                Logger.debug(Strings.receipt.throttling_force_refreshing_receipt)
+
+                // If requested to refresh again within the throttle duration
+                // Use `ReceiptRefreshPolicy.onlyIfEmpty` so receipt is not refreshed if it's already loaded.
+                self.receiptData(refreshPolicy: .onlyIfEmpty, completion: completion)
+            } else {
+                Logger.debug(Strings.receipt.force_refreshing_receipt)
+                self.refreshReceipt(completion)
+            }
```

Comment:
makes sense. I wonder if a more reliable way of doing this would be to throttle unless we've had a response for the last request, i.e.: there are no requests in flight. 

Do we know from the logs if for the customers that encountered this, they were throttled while a refresh was in flight or they were throttled after successful but frequent requests finished?

Reply:
> makes sense. I wonder if a more reliable way of doing this would be to throttle unless we've had a response for the last request, i.e.: there are no requests in flight.

In most of these cases we're not doing concurrent requests, because we handle transactions linearly (see the example log in #2116). The problem is that we're refreshing it multiple times every second.

> Do we know from the logs if for the customers that encountered this, they were throttled while a refresh was in flight or they were throttled after successful but frequent requests finished?

See log:
```
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218313433 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218320027 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218327443 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218334901 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218342186 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218353996 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218362596 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218370255 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
DEBUG: ℹ️ PaymentQueue updatedTransaction: product2 2000000218377794 2000000042575785 1
DEBUG: ℹ️ Force refreshing the receipt to get latest transactions from Apple.
<SKReceiptRefreshRequest: 0x283edc860>: Finished refreshing receipt with error: Error Domain=ASDErrorDomain Code=603 "Request throttled" UserInfo={NSLocalizedFailureReason=Unified receipt is valid and current, NSLocalizedDescription=Request throttled, AMSServerErrorCode=0}
DEBUG: ℹ️ Loaded receipt from url file:///private/var/mobile/Containers/Data/Application/5E49D650-1DE9-40C7-A20C-147D6C8718F8/StoreKit/sandboxReceipt
```

All those happen within the same second.

Reply:
that highlights what I mean, though - we don't wait until we get the response from SKReceiptRefreshRequest before allowing for the next request to start, right? the logs there show a single finish

Reply:
Ooooh I see what you're saying.
Okay, I think maybe a better fix would be to reuse the completion block for concurrent receipt requests instead of throttling them.

That's what you have in mind too?

Reply:
chatted about this offline - we already have a mechanism for this, it's just not enough if the receipt request finishes quickly.